samples/cpp/intelligent_scissors.cpp

```diff
@@ -0,0 +1,210 @@
+                        {
+                            cost_map.at<float>(q) = cost;
+                            hit_map_x.at<int>(q)= p.x;
+                            hit_map_y.at<int>(q) = p.y;
+                            processed.at<uchar>(q) = 1;
+                            Pix val;
+                            val.cost = cost_map.at<float>(q);
+                            val.next_point = q;
+                            L.push(val);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+static void onMouse(int event, int x, int y, int , void *)
+{
+    if (event == EVENT_LBUTTONDOWN)
+    {
+        end = Point(x, y);
+        if (!contours.back().empty())
+        {
+            for (int i = static_cast<int>(tmp_contour.size()) - 1; i >= 0; i--)
+            {
+                contours.back().push_back(tmp_contour[i]);
+            }
+            tmp_contour.clear();
+        }
+        else
+        {
+            contours.back().push_back(end);
+        }
+        find_min_path(end);
+
+        img_render.copyTo(img_pre_render);
+        imshow("lasso", img_render);
+    }
+    else if (event == EVENT_RBUTTONDOWN)
+    {
+        img_pre_render.copyTo(img_render);
+        drawContours(img_pre_render, contours, static_cast<int>(contours.size()) - 1, Scalar(0,255,0), FILLED);
+        addWeighted(img_pre_render, 0.3, img_render, 0.7, 0, img_render);
+        contours.resize(contours.size() + 1);
+        imshow("lasso", img_render);
+    }
+    else if (event == EVENT_MOUSEMOVE && !contours.back().empty())
+    {
+        tmp_contour.clear();
+        img_pre_render.copyTo(img_render);
+        Point val_point = Point(x, y);
+        while (val_point != end)
+        {
+            tmp_contour.push_back(val_point);
+            Point cur = Point(hit_map_x.at<int>(val_point), hit_map_y.at<int>(val_point));
+            line(img_render, val_point, cur, Scalar(255, 0, 0), 2);
+            val_point = cur;
+        }
+        imshow("lasso", img_render);
+    }
+}
+
+const char* keys =
+{
+    "{help h | |}{@image|fruits.jpg|}"
+};
+
+int main( int argc, const char** argv )
+{
+    CommandLineParser parser(argc, argv, keys);
+    parser.about("\nThis program demonstrates implementation of 'intelligent scissors' algorithm\n"
+                 "To start drawing a new contour select a pixel, click LEFT mouse button.\n"
+                 "To fix a path click LEFT mouse button again.\n"
+                 "To finish drawing a contour click RIGHT mouse button.\n");
+    if (parser.has("help"))
+    {
+        parser.printMessage();
+        return 1;
+    }
+
+    std::string filename = parser.get<std::string>(0);
+
+    Mat grayscale, img_canny;
+    img = imread(samples::findFile(filename));
+    hit_map_x.create(img.rows, img.cols, CV_32SC1);
+    hit_map_y.create(img.rows, img.cols, CV_32SC1);
+
+    cvtColor(img, grayscale, COLOR_BGR2GRAY);
+    Canny(grayscale, img_canny, 50, 100);
```

Comment:
> 50, 100

Magic numbers.

```
int EDGE_THRESHOLD_LOW = 50;
int EDGE_THRESHOLD_HIGH = 100;
```

Also they can be defined thorough command line parameters.